app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt

```diff
@@ -0,0 +1,209 @@
+        _products.value = newProducts
+        search()
+    }
+
+    val isDeparture: LiveData<Boolean> = _isDeparture
+
+    fun setIsDeparture(departure: Boolean) {
+        _isDeparture.value = departure
+        search()
+    }
+
+    val isExpanded: LiveData<Boolean> = _isExpanded
+
+    fun setIsExpanded(expanded: Boolean) {
+        _isExpanded.value = expanded
+    }
+
+    fun toggleIsExpanded() {
+        _isExpanded.value = !_isExpanded.value!!
+    }
+
+    val isFavTrip: MutableLiveData<Boolean>
+        get() = _tripsRepository.isFavTrip
+
+    fun toggleFavTrip() {
+        _tripsRepository.toggleFavState()
+    }
+
+    override fun onLocationItemClick(loc: WrapLocation, type: FavLocationType) {
+        when (type) {
+            FavLocationType.FROM -> setFromLocation(loc)
+            FavLocationType.VIA -> setViaLocation(loc)
+            FavLocationType.TO -> setToLocation(loc)
+        }
+        search()
+        // clear finding GPS location request
+        if (findGpsLocation.value == type) findGpsLocation.value = null
+    }
+
+    override fun onLocationCleared(type: FavLocationType) {
+        when (type) {
+            FavLocationType.FROM -> setFromLocation(null)
+            FavLocationType.VIA -> {
+                setViaLocation(null)
+                search()
+            }
+            FavLocationType.TO -> setToLocation(null)
+        }
+        // clear finding GPS location request
+        if (findGpsLocation.value == type) findGpsLocation.value = null
+    }
+
+    /* Trip Queries */
+    fun search() {
+        val from = _fromLocation.value; val to = _toLocation.value
+        val via = if (_isExpanded.value != null && _isExpanded.value!!)
+            _viaLocation.value else null
+        val calendar = if (_now.value != null && _now.value!!)
+            Calendar.getInstance() else _calendar.value
+        if (from == null || to == null || calendar == null) return
+        _calendar.value = calendar
+
+        val tripQuery = TripQuery(from, via, to, calendar.time, _isDeparture.value, _products.value)
+        _tripsRepository.search(tripQuery)
+        showTrips.call()
+    }
+
+    fun searchMore(later: Boolean) {
+        _tripsRepository.searchMore(later)
+    }
+
+    val queryMoreState: LiveData<QueryMoreState>
+        get() = _tripsRepository.queryMoreState
+    val trips: LiveData<Set<Trip>>
+        get() = _tripsRepository.trips
+    val queryError: LiveData<String>
+        get() = _tripsRepository.queryError
+    val queryPTEError: LiveData<Pair<String, String>>
+        get() = _tripsRepository.queryPTEError
+    val queryMoreError: LiveData<String>
+        get() = _tripsRepository.queryMoreError
+
+    init {
+        _now.value = true
+        _calendar.value = Calendar.getInstance()
+        _products.value = EnumSet.allOf(Product::class.java)
+        _isDeparture.value = true
+        _isExpanded.value = false
+        topSwipeEnabled.value = false
```

Comment:
You can set these initial values in the live data constructor instead.

Reply:
True, might be cleaner.